src/internal/watch.ts

```diff
@@ -11,30 +11,52 @@
 //  handlePropChange(oldValue, newValue) {
 //    ...
 //  }
-//
+
+import type { LitElement } from 'lit';
+import type { NonUndefined } from 'utility-types';
+
+type UpdateHandler = (prev?: unknown, next?: unknown) => void;
+
+// Mostly copied from utility-types' FunctionKeys type but this uses a specific function signature
+type UpdateHandlerFunctionKeys<T extends object> = {
+  [K in keyof T]-?: NonUndefined<T[K]> extends UpdateHandler ? K : never;
+}[keyof T];
+
 interface WatchOptions {
+  /**
+   * If true, will only start watching after the initial update/render
+   */
   waitUntilFirstUpdate?: boolean;
 }
 
 export function watch(propName: string, options?: WatchOptions) {
-  return (protoOrDescriptor: any, name: string): any => {
-    const { update } = protoOrDescriptor;
-
-    options = Object.assign({ waitUntilFirstUpdate: false }, options) as WatchOptions;
-
-    protoOrDescriptor.update = function (changedProps: Map<string, any>) {
-      if (changedProps.has(propName)) {
-        const oldValue = changedProps.get(propName);
-        const newValue = this[propName];
+  const resolvedOptions: Required<WatchOptions> = {
+    waitUntilFirstUpdate: false,
+    ...options
+  };
+  return <ElemClass extends LitElement>(
+    proto: ElemClass,
+    decoratedFnName: UpdateHandlerFunctionKeys<ElemClass>
+  ): void => {
+    // @ts-expect-error -- update is a protected property
+    const { update } = proto;
+    if (propName in proto) {
+      const propNameKey = propName as keyof ElemClass;
+      // @ts-expect-error -- update is a protected property
+      proto.update = function (this: ElemClass, changedProps: Map<keyof ElemClass, ElemClass[keyof ElemClass]>) {
+        if (changedProps.has(propNameKey)) {
+          const oldValue = changedProps.get(propNameKey);
+          const newValue = this[propNameKey];
 
-        if (oldValue !== newValue) {
-          if (!options?.waitUntilFirstUpdate || this.hasUpdated) {
-            this[name].call(this, oldValue, newValue);
+          if (oldValue !== newValue) {
+            if (!resolvedOptions.waitUntilFirstUpdate || this.hasUpdated) {
+              (this[decoratedFnName] as unknown as UpdateHandler)(oldValue, newValue);
+            }
           }
         }
-      }
 
-      update.call(this, changedProps);
-    };
+        update.call(this, changedProps);
+      };
+    }
```

Comment:
This is one instance where making TypeScript happy may not be worth the effort and added dependency. I didn't notice `utility-types` being used anywhere else (but then again, I wasn't looking for it). The new code looks significantly more complex than the old code, and for its intended usage it's probably OK to live with.

This could just as easily be a standalone decorator. I can live with it if you'd rather leave it in, but complex typings make me a bit wary as a maintainer.

Reply:
Yeah we could probably just in-line the few pieces of `utility-types` to remove the dependency. Good point. And let me know what you want to do here. It does have the added benefit of yelling at you if you try to decorate a method which is incompatible, which is really nice. But yeah this function is much more gnarly as a result.

Reply:
Removing the dep is a good compromise. Since the work is already done, might as well keep the types.